classes/plugins/ThemePlugin.inc.php

```diff
@@ -275,13 +329,230 @@ public function &getScript($name) {
+
+		if (isset($this->options[$name])) {
+			return $this->options[$name];
+		}
+
+		return $this->parent ? $this->parent->getOptionConfig($name) : false;
+	}
+
+	/**
+	 * Get all options' configuration settings.
+	 *
+	 * This retrieves a single array containing options settings for this
+	 * theme and any parent themes.
+	 *
+	 * @return array
+	 */
+	public function getOptionsConfig() {
+
+		if (!$this->parent) {
+			return $this->options;
+		}
+
+		return array_merge(
+			$this->parent->getOptionsConfig(),
+			$this->options
+		);
+	}
+
+	/**
+	 * Get all option values
+	 *
+	 * This retrieves a single array containing option values for this theme
+	 * and any parent themes.
+	 *
+	 * @return array
+	 */
+	public function getOptionsValues() {
+
+		$pluginSettingsDAO = DAORegistry::getDAO('PluginSettingsDAO');
+
+		$context = Request::getContext();
+		$contextId = empty($context) ? 0 : $context->getId();
+		$values = $pluginSettingsDAO->getPluginSettings($contextId, $this->getName());
+
+		if (!$this->parent) {
+			return $values;
+		}
+
+		return array_merge(
+			$this->parent->getOptionsValues(),
+			$values
+		);
+	}
+
+	/**
+	 * Sanitize and save a theme option
+	 *
+	 * @param $name string A unique id for the option to save
+	 * @param $value mixed The new value to save
+	 * @param $contextId int Optional context id. Defaults to the current
+	 *  context
+	 */
+	public function saveOption($name, $value, $contextId = null) {
+
+		$option = !empty($this->options[$name]) ? $this->options[$name] : null;
+
+		if (is_null($option)) {
+			return $this->parent ? $this->parent->saveOption($name, $value) : false;
+		}
+
+		$type = '';
+		switch ($option['type']) {
+			case 'text' :
+            case 'select' :
+			case 'color' :
+				$type = 'text';
+				break;
+		}
+
+		if (is_null($contextId)) {
+			$context = Request::getContext();
+			$contextId = $context->getId();
+		}
+
+		$this->updatesetting($contextId, $name, $value, $type);
+
+		// Clear the template cache so that new settings can take effect
+		$templateMgr = TemplateManager::getManager($this->getRequest());
+		$templateMgr->clearTemplateCache();
+		$templateMgr->clearCssCache();
+	}
+
+	/**
+	 * Save options in any form
+	 *
+	 * This helper function allows you to save theme options attached to any
+	 * form by hooking into the form's execute function.
+	 *
+	 * @see Form::execute()
+	 * @param $hookName string
+	 * @param $args array Arguments passed via the hook
+	 *  `form` Form The form object from which option values can be retrieved.
+	 *  `request` Request
+	 */
+	public function saveOptionsForm($hookName, $args) {
+
+		$form = $args[0];
+
+		$options = $this->getOptionsConfig();
+
+		// Ensure theme options from the site-wide settings form are applied
+		// to the site-wide context
+		if ($hookName == 'sitesetupform::execute') {
+			$contextId = 0;
+		}
+
+		foreach ($options as $optionName => $optionArgs) {
+			$value = $form->getData('themeOption_' . $optionName);
```

Comment:
It might be worth `define`ing the `themeOption_` prefix once, then using that throughout.